### PR TITLE
[XDK][NDK] Improve and sync KUSER_SHARED_DATA

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Sync KUSER_SHARED_DATA between NDK and XDK.

## Proposed changes

- Add missing XSTATE definitions to NDK
- Update XSTATE stuff to latest SDK
- Apply our standard formatting
- Some versioning improvements
- Add field offsets
- Sync KUSER_SHARED_DATA between NDK and XDK

## TODO

Decide upon how to handle the (growing) number of `#if (NTDDI_VERSION ...)`. A lot of the layout is compatible between versions and adding more fields doesn't change anything. To enable support for modern features like XSTATE configuration, it will be needed to add `|| defined(__REACTOS__)`. Please provide feedback.

- [ ] Keep them all
- [ ] Remove them for newer fields that replace padding
- [ ] Remove them for newer fields that replace Reserved fields
- [ ] Remove them, unless required: use the more modern definition, unless we already use the old one in the kernel.

### CODE REVIEW

File: sdk/include/ndk/ketypes.h [hunk 1]
```diff
@@ -518,78 +518,171 @@ typedef struct _KSYSTEM_TIME
...
+            ULONG OptimizedSave:1;
+            ULONG CompactionEnabled:1; // WIN10+
+        };
+    };
+    XSTATE_FEATURE Features[MAXIMUM_XSTATE_FEATURES];
+#if (NTDDI_VERSION >= NTDDI_WIN10)
+    ULONG64 EnabledSupervisorFeatures;
+    ULONG64 AlignedFeatures;
+    ULONG AllFeatureSize;
+    ULONG AllFeatures[MAXIMUM_XSTATE_FEATURES];
```
Comment (JoachimHenze) [inline sdk/include/ndk/ketypes.h:550]:
```suggestion
    ULONG AllFeatures[MAXIMUM_XSTATE_FEATURES];
```
EOL-whitespace here. I do recommend to (not just accept my suggestion), but to strip all of them in that file via NPP++.
I am not sure that I catched all. Just spotted that one by chance.

**Review by JoachimHenze — CHANGES_REQUESTED** [review]
EOL whitespaces should be stripped (everywhere).

File: sdk/include/ndk/ketypes.h [hunk 2]
```diff
@@ -518,78 +518,171 @@ typedef struct _KSYSTEM_TIME
...
+#endif
+    ULONG ImageFileExecutionOptions;                        // 0x3a0
+#if (NTDDI_VERSION >= NTDDI_VISTASP1)
+    ULONG LangGenerationCount;                              // 0x3a4
+#else
+  /* 4 bytes padding */
+#endif
+    ULONGLONG Reserved5;                                    // 0x3a8
+    volatile ULONG64 InterruptTimeBias;                     // 0x3b0
+#endif
```
Comment (HBelusca) [inline sdk/include/ndk/ketypes.h:674]:
```suggestion
#endif /* NTDDI_VERSION >= NTDDI_VISTA */
```

File: sdk/include/ndk/ketypes.h [hunk 3]
```diff
@@ -518,78 +518,171 @@ typedef struct _KSYSTEM_TIME
...
+    };
+    XSTATE_FEATURE Features[MAXIMUM_XSTATE_FEATURES];
+#if (NTDDI_VERSION >= NTDDI_WIN10)
+    ULONG64 EnabledSupervisorFeatures;
+    ULONG64 AlignedFeatures;
+    ULONG AllFeatureSize;
+    ULONG AllFeatures[MAXIMUM_XSTATE_FEATURES];    
+#endif
+#if (NTDDI_VERSION >= NTDDI_WIN10_RS5)
+    ULONG64 EnabledUserVisibleSupervisorFeatures;
```
Comment (HBelusca) [inline sdk/include/ndk/ketypes.h:553]: For example that one has a trailing space

File: sdk/include/ndk/ketypes.h [hunk 4]
```diff
@@ -518,78 +518,171 @@ typedef struct _KSYSTEM_TIME
     LONG High2Time;
 } KSYSTEM_TIME, *PKSYSTEM_TIME;
 
+#define MAXIMUM_XSTATE_FEATURES             64
+
+typedef struct _XSTATE_FEATURE
+{
+  ULONG Offset;
+  ULONG Size;
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply sdk/include/ndk/ketypes.h:526]: Copy-paste. I'll fix it.

File: sdk/include/ndk/ketypes.h (hunk 1 above)
Comment (tkreuzer) [reply sdk/include/ndk/ketypes.h]: Fixed

**Review by JoachimHenze — APPROVED** [review]
Approved without deep understanding of the code. My formal concerns have been addressed properly. I would appreciate mentioning in the description what those _XSTATE are used for and whether you stumbled across the mismatch only in x64 context or whether that matters also for x86. Thx.

**Comment by tkreuzer** [comment]
> I would appreciate mentioning in the description what those _XSTATE are used for and whether you stumbled across the mismatch only in x64 context or whether that matters also for x86. Thx.

XSTATE data is used for saving/restoring extended context state like AVL registers. The changes matter for both x86 and x64.

**Review by HBelusca — APPROVED** [review]
Current version of the PR is OK

**Pull request merged**